components/ai_chat/renderer/page_content_extractor.cc

```diff
@@ -0,0 +1,223 @@
+#include "content/public/renderer/render_frame.h"
+#include "content/public/renderer/render_frame_observer.h"
+#include "net/base/url_util.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "third_party/blink/public/platform/web_string.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+#include "url/gurl.h"
+#include "url/url_constants.h"
+#include "v8/include/v8-isolate.h"
+
+namespace {
+
+const char16_t kYoutubeTranscriptUrlExtractionScript[] =
+    uR"JS(
+      (function() {
+        return ytplayer?.config?.args?.raw_player_response?.captions?.playerCaptionsTracklistRenderer?.captionTracks?.[0]?.baseUrl
+      })()
+    )JS";
+
+const char16_t kVideoTrackTranscriptUrlExtractionScript[] =
+    // TODO(petemill): Make more informed srclang choice.
+    // TODO(petemill): Observe <video>.textTracks
+    uR"JS(
+      (function() {
+        const nodes = document.querySelectorAll('video track')
+        if (nodes.length) {
+          let selectedNode = nodes[0]
+          for (const node of nodes) {
+            if (node.srclang.toLowerCase() === 'en') {
+              selectedNode = node
+            }
+          }
+          return node.src
+        }
+      })()
+    )JS";
+
+constexpr auto kYouTubeHosts = base::MakeFixedFlatSet<base::StringPiece>(
+    {"youtube.com", "www.youtube.com", "m.youtube.com"});
+
+// TODO(petemill): Use heuristics to determine if page's main focus is
+// a video, and not a hard-coded list of Url hosts.
+constexpr auto kVideoTrackHosts =
+    base::MakeFixedFlatSet<base::StringPiece>({"www.ted.com"});
+
+}  // namespace
+
+namespace ai_chat {
+
+PageContentExtractor::PageContentExtractor(
+    content::RenderFrame* render_frame,
+    service_manager::BinderRegistry* registry,
+    int32_t global_world_id,
+    int32_t isolated_world_id)
+    : content::RenderFrameObserver(render_frame),
+      global_world_id_(global_world_id),
+      isolated_world_id_(isolated_world_id),
+      weak_ptr_factory_(this) {
+  if (!render_frame->IsMainFrame()) {
+    return;
+  }
+  // Bind mojom API to allow browser to communicate with this class
+  // Being a RenderFrameObserver, this object is scoped to the RenderFrame.
+  // Unretained is safe here because `registry` is also scoped to the
+  // RenderFrame.
+  registry->AddInterface(base::BindRepeating(
+      &PageContentExtractor::BindReceiver, base::Unretained(this)));
+}
+
+PageContentExtractor::~PageContentExtractor() = default;
+
+void PageContentExtractor::OnDestruct() {
+  delete this;
+}
+
+void PageContentExtractor::ExtractPageContent(
+    mojom::PageContentExtractor::ExtractPageContentCallback callback) {
+  VLOG(1) << "AI Chat renderer has been asked for page content.";
+  blink::WebLocalFrame* main_frame = render_frame()->GetWebFrame();
+  GURL origin =
+      url::Origin(((const blink::WebFrame*)main_frame)->GetSecurityOrigin())
+          .GetURL();
```

Comment:
As this code is part of renderer, it'd be better to use blink::WebSecurityOrigin as is rather than GURL.

```suggestion
auto security_origin = const_cast<const blink::WebLocalFrame*>(main_frame)->GetSecurityOrigin();

security_origin.IsSameOriginWith(blink::WebSecurityOrigin::Create("youtube.com"));  // not sure but this might cover www.youtube.com, m.youtube, youtube.com, youtube.co.uk/kr and etc together.
```

Reply:
Actually it seems that
1. YouTube always redirects to either `m.youtube.com` or `www.youtube.com`. I do not see any other variations.
2. `IsSameOriginWith` compares exact host, will not return `true` if the host has different endings or prefixes.

For those reasons, I'd like to keep the `if (base::Contains(kYouTubeHosts, host))` logic I have for now. That ok? Maybe I'll remove the `"youtube.com"` one.